monitor/cert_submitter.go

```diff
@@ -0,0 +1,143 @@
+package monitor
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/x509"
+	"fmt"
+	"log"
+	"time"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/jmhodges/clock"
+	"github.com/letsencrypt/ct-woodpecker/pki"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// certSubmitterStats is a type to hold the prometheus metrics used by
+// a certSubmitter
+type certSubmitterStats struct {
+	certSubmitLatency *prometheus.HistogramVec
+	certSubmitResults *prometheus.CounterVec
+}
+
+var (
+	// submitTimeout controls how long each certificate chain submission should
+	// wait before timing out
+	submitTimeout = time.Second * 15
+	// requiredSCTFreshness indicates how fresh a timestamp in a returned SCT must
+	// be for it to be considered valid.
+	requiredSCTFreshness = time.Minute * 10
+
+	// certStats is a certSubmitterStats instance with promauto registered
+	// prometheus metrics
+	certStats *certSubmitterStats = &certSubmitterStats{
+		certSubmitLatency: promauto.NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "cert_submit_latency",
+			Help:    "Latency submitting certificate chains to CT logs",
+			Buckets: internetFacingBuckets,
+		}, []string{"uri"}),
+		certSubmitResults: promauto.NewCounterVec(prometheus.CounterOpts{
+			Name: "cert_submit_results",
+			Help: "Count of results from submitting certificate chains to CT logs, sliced by status",
+		}, []string{"uri", "status"}),
+	}
+)
+
+// certSubmitter is a type for periodically issuing certificates and submitting
+// them to a log's add-chain endpoint.
+type certSubmitter struct {
+	logger *log.Logger
+	clk    clock.Clock
+	client monitorCTClient
+	logURI string
+	stats  *certSubmitterStats
+
+	// How long to sleep between submitting certificates to the log
+	certSubmitInterval time.Duration
+	// ECDSA private key used to issue certificates to submit to the log. Nil if
+	// no certificates are to be submitted to the log.
```

Comment:
"Nil if no certificates are to be submitted": Reading the code in monitor.go, it looks like `certSubmitter` won't be constructed if `certIssuerKey` is nil. Want to just document these two fields as mandatory? In other words, don't construct a certSubmitter if you don't want to submit.